src/useKeyPress.ts

```diff
@@ -0,0 +1,63 @@
+import * as React from "react";
+const { useState, useEffect } = React;
+
+interface Options {
+  useKeyboardJS: boolean;
+}
+
+const defaults: Options = {
+  useKeyboardJS: false
+};
+
+const useKeyPress = (targetKey: string, config: Options = defaults) => {
+  const [state, setState] = useState(false);
+  const { useKeyboardJS } = config;
+
+  let keyboardjs;
+
+  if (useKeyboardJS) {
+    import("keyboardjs").then(module => {
+      keyboardjs = module;
+    });
+  }
+
+  const regularDownHandler = ({ key }: KeyboardEvent) => {
+    if (key === targetKey) {
+      setState(true);
+    }
+  };
+
+  const regularUpHandler = ({ key }: KeyboardEvent) => {
+    if (key === targetKey) {
+      setState(false);
+    }
+  };
+
+  const customDownHandler = () => {
+    setState(true);
+  };
+  const customUpHandler = () => {
+    setState(false);
+  };
+
+  useEffect(() => {
+    if (useKeyboardJS) {
+      keyboardjs.bind(targetKey, customDownHandler, customUpHandler);
+    } else {
+      window.addEventListener("keydown", regularDownHandler);
+      window.addEventListener("keyup", regularUpHandler);
+    }
+    return () => {
+      if (useKeyboardJS) {
+        keyboardjs.unbind(targetKey, customDownHandler, customUpHandler);
+      } else {
+        window.removeEventListener("keydown", regularDownHandler);
+        window.removeEventListener("keyup", regularUpHandler);
+      }
+    };
+  }, []);
```

Comment:
The `[]` should probably be

```js
[targetKey, useKeyboardJS]
```

Reply:
Sure! My bad this perf thing when not skipping these. I better do `[targetKey, config]`, don't you think so?

Reply:
If `useEffect` supports an object, then yes.

Reply:
`...Object.values(config)` does not guarantee the order of the values, right? But the order is important here.

Reply:
According to the spec, [`Object.values`](https://read262.netlify.com/fundamental-objects/object-objects#sec-object.values) calls [`EnumerableOwnPropertyNames`](https://read262.netlify.com/abstract-operations/operations-on-objects#sec-enumerableownpropertynames), which reads [`[[OwnPropertyKeys]]`](https://read262.netlify.com/ordinary-and-exotic-objects-behaviours/ordinary-object-internal-methods-and-internal-slots#sec-ordinary-object-internal-methods-and-internal-slots-ownpropertykeys), which calls [`OrdinaryOwnPropertyKeys`](https://read262.netlify.com/ordinary-and-exotic-objects-behaviours/ordinary-object-internal-methods-and-internal-slots#sec-ordinaryownpropertykeys), which specifies an order for the returned array.

Reply:
I think the shallow comparison thing could be solved using `useRef` as suggested [here](https://github.com/kentcdodds/react-github-profile/blob/egghead-2018/refactor-03/src/screens/user/components/query.js#L19):

```ts
  useEffect(() => {
    if (isEqual(previousInputs.current, [targetKey, config])) {
      return;
    }

    if (useKeyboardJS) {
      keyboardjs.bind(targetKey, customDownHandler, customUpHandler);
    } else {
      window.addEventListener("keydown", regularDownHandler);
      window.addEventListener("keyup", regularUpHandler);
    }
    return () => {
      if (useKeyboardJS) {
        keyboardjs.unbind(targetKey, customDownHandler, customUpHandler);
      } else {
        window.removeEventListener("keydown", regularDownHandler);
        window.removeEventListener("keyup", regularUpHandler);
      }
    };
  });

  const previousInputs = useRef(null as unknown);
  useEffect(() => {
    previousInputs.current = [targetKey, config];
  });
``` 

But I don't really think it results in the expected behaviour - neither does `[targetKey, useKeyboardJS]`. 

I was reconsidering and I think that, in this case, the second argument to `useEffect` should actually be an empty array. Could you elaborate more on your previous suggestion, @streamich ?

Reply:
What do you mean `[targetKey, useKeyboardJS]` does not result in expected behaviour?

I think you should either use that or sort the keys manually:

```js
[targetKey, ...config.entries().sort(([k1], [k2]) => k1 < k2 ? 1 : -1).map(([_, value]) => value)]
```

Reply:
nm, @streamich; my bad. 

Anyway, added and it seems to be okay. Could you check again and tell me if there's anything else missing in the PR?

Reply:
Looks good.